h/streamer.py

```diff
@@ -655,12 +620,82 @@ def should_send_annotation_event(socket, annotation, event_data):
     return True
 
 
+def process_message(reader, message):
+    """
+    Process a message from the queue.
+
+    Deserializes and distributes messages from the queue to the appropriate
+    handler functions.
+    """
+    if reader.topic == ANNOTATIONS_TOPIC:
+        broadcast_annotation_message(message, WebSocket.instances)
+    elif reader.topic == USER_TOPIC:
+        broadcast_session_change_message(message, WebSocket.instances)
+
+
+def process_queue(settings, topics):
+    """
+    Configure, start, and monitor the queue readers.
+
+    This sets up all the passed readers to route their messages to
+    `process_message`, starts them, and then enters a loop that checks that the
+    reader threads never join. If they do, this fact is logged and all readers
+    are killed.
+    """
+    channel = 'stream-{}#ephemeral'.format(_random_id())
+    readers = [queue.get_reader(settings, topic, channel)
+               for topic in topics]
+
+    # Start all the readers
+    for reader in readers:
+        reader.on_message.connect(receiver=process_message)
+        reader.start(block=False)
+
+    _wait_for_readers(readers)
+
+    # We should never get here. If we do, it's because a reader thread has
+    # prematurely quit.
+    log.error("queue reader exited: killing readers and starting again")
+    for reader in readers:
+        reader.close()
+
+
 def _random_id():
     """Generate a short random string"""
     data = struct.pack('Q', random.getrandbits(64))
     return base64.urlsafe_b64encode(data).strip(b'=')
 
 
+def _wait_for_readers(readers):
+    """Wait for one or more of the readers to exit."""
+    while 1:
```

Comment:
Meh. Once upon a time I was told that `while 1:` was more Pythonic (for what are probably highly dubious reasons... did you know you can reassign `True`.... `True = 0`?), but tbh you're probably right. I'll change it.